From 1e067c3e5c06ff063affbc6f3df07af279700518 Mon Sep 17 00:00:00 2001
From: Til Kaiser <mail@tk154.de>
Date: Tue, 14 Apr 2026 13:39:08 +0200
Subject: [PATCH 2/3] filter: add bpf_xdp_to_sgvec helper

---
 include/linux/filter.h |  8 ++++++
 net/core/filter.c      | 55 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 63 insertions(+)

diff --git a/include/linux/filter.h b/include/linux/filter.h
index a91f2babf425..0d4eef6ae734 100644
--- a/include/linux/filter.h
+++ b/include/linux/filter.h
@@ -1772,6 +1772,8 @@ int __bpf_xdp_store_bytes(struct xdp_buff *xdp, u32 offset, void *buf, u32 len);
 void *bpf_xdp_pointer(struct xdp_buff *xdp, u32 offset, u32 len);
 void bpf_xdp_copy_buf(struct xdp_buff *xdp, unsigned long off,
 		      void *buf, unsigned long len, bool flush);
+int bpf_xdp_to_sgvec(struct xdp_buff *xdp, struct scatterlist *sg,
+			  u32 offset, u32 len);
 #else /* CONFIG_NET */
 static inline int __bpf_skb_load_bytes(const struct sk_buff *skb, u32 offset,
 				       void *to, u32 len)
@@ -1806,6 +1808,12 @@ static inline void bpf_xdp_copy_buf(struct xdp_buff *xdp, unsigned long off, voi
 				    unsigned long len, bool flush)
 {
 }
+
+static inline int bpf_xdp_to_sgvec(struct xdp_buff *xdp, struct scatterlist *sg,
+					u32 offset, u32 len)
+{
+	return -EOPNOTSUPP;
+}
 #endif /* CONFIG_NET */
 
 #endif /* __LINUX_FILTER_H__ */
diff --git a/net/core/filter.c b/net/core/filter.c
index 560ea52f331a..8b9f341afa67 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -4085,6 +4085,61 @@ void *bpf_xdp_pointer(struct xdp_buff *xdp, u32 offset, u32 len)
 	return offset + len <= size ? addr + offset : NULL;
 }
 
+int bpf_xdp_to_sgvec(struct xdp_buff *xdp, struct scatterlist *sg,
+               	   	 u32 offset, u32 len)
+{
+    struct skb_shared_info *sinfo = xdp_get_shared_info_from_buff(xdp);
+    int start = xdp->data_end - xdp->data;
+    int i, copy = start - offset;
+    int elt = 0;
+
+    /* Linear part in xdp->data..data_end */
+    if (copy > 0) {
+        if (copy > len)
+            copy = len;
+        sg_set_buf(sg, xdp->data + offset, copy);
+        elt++;
+        if ((len -= copy) == 0)
+            goto out;
+        offset += copy;
+    }
+
+	if (!xdp_buff_has_frags(xdp))
+		goto out;
+
+    /* Page frags from xdp_shared_info */
+    for (i = 0; i < sinfo->nr_frags; i++) {
+        int end;
+
+		WARN_ON(start > offset + len);
+
+		end = start + skb_frag_size(&sinfo->frags[i]);
+        if ((copy = end - offset) > 0) {
+            skb_frag_t *frag = &sinfo->frags[i];
+			if (unlikely(elt && sg_is_last(&sg[elt - 1])))
+				return -EMSGSIZE;
+
+            if (copy > len)
+                copy = len;
+            sg_set_page(&sg[elt], skb_frag_page(frag), copy,
+                        skb_frag_off(frag) + offset - start);
+            elt++;
+            if (!(len -= copy))
+                goto out;
+            offset += copy;
+        }
+        start = end;
+    }
+
+out:
+	if (elt)
+    	sg_mark_end(&sg[elt - 1]);
+
+    BUG_ON(len);
+    return elt;
+}
+EXPORT_SYMBOL_GPL(bpf_xdp_to_sgvec);
+
 BPF_CALL_4(bpf_xdp_load_bytes, struct xdp_buff *, xdp, u32, offset,
 	   void *, buf, u32, len)
 {
-- 
2.53.0

