From b9688fa67515c890fc57a370c652bcc63785ae3b Mon Sep 17 00:00:00 2001
From: Til Kaiser <mail@tk154.de>
Date: Wed, 14 Jan 2026 21:44:10 +0100
Subject: [PATCH 2/3] bpf: add kfuncs to checksum XDP and SKB packet data

---
 net/core/filter.c | 73 +++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 73 insertions(+)

diff --git a/net/core/filter.c b/net/core/filter.c
index fb159fd5beeb..478bd156f552 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -12159,6 +12159,77 @@ __bpf_kfunc int bpf_sk_assign_tcp_reqsk(struct __sk_buff *s, struct sock *sk,
 #endif
 }
 
+/* Checksum xdp data. */
+__bpf_kfunc u32 bpf_xdp_checksum(struct xdp_md *xdp_ctx, u32 offset, u32 len, u32 csum)
+{
+	struct xdp_buff *xdp = (struct xdp_buff *)xdp_ctx;
+	int start = xdp->data_end - xdp->data;
+	struct skb_shared_info *sinfo;
+	int i, copy = start - offset;
+	int pos = 0;
+
+	/* Checksum header. */
+	if (copy > 0) {
+		if (copy > len)
+			copy = len;
+		csum = csum_partial(xdp->data + offset, copy, csum);
+		if ((len -= copy) == 0)
+			return csum;
+		offset += copy;
+		pos	= copy;
+	}
+
+	if (likely(!xdp_buff_has_frags(xdp)))
+		goto out;
+
+	sinfo = xdp_get_shared_info_from_buff(xdp);
+
+	for (i = 0; i < sinfo->nr_frags; i++) {
+		int end;
+		skb_frag_t *frag = &sinfo->frags[i];
+
+		WARN_ON(start > offset + len);
+
+		end = start + skb_frag_size(frag);
+		if ((copy = end - offset) > 0) {
+			u32 p_off, p_len, copied;
+			struct page *p;
+			__wsum csum2;
+			u8 *vaddr;
+
+			if (copy > len)
+				copy = len;
+
+			skb_frag_foreach_page(frag,
+					      skb_frag_off(frag) + offset - start,
+					      copy, p, p_off, p_len, copied) {
+				vaddr = kmap_atomic(p);
+				csum2 = csum_partial(vaddr + p_off, p_len, 0);
+				kunmap_atomic(vaddr);
+				csum = csum_block_add(csum, csum2, pos);
+				pos += p_len;
+			}
+
+			if (!(len -= copy))
+				return csum;
+			offset += copy;
+		}
+		start = end;
+	}
+
+out:
+	BUG_ON(len);
+
+	return csum;
+}
+
+__bpf_kfunc u32 bpf_skb_checksum(struct __sk_buff *skb_ctx, u32 offset, u32 len, u32 csum)
+{
+	struct sk_buff *skb = (struct sk_buff *)skb_ctx;
+
+	return skb_checksum(skb, offset, len, csum);
+}
+
 __bpf_kfunc_end_defs();
 
 int bpf_dynptr_from_skb_rdonly(struct __sk_buff *skb, u64 flags,
@@ -12178,10 +12249,12 @@ int bpf_dynptr_from_skb_rdonly(struct __sk_buff *skb, u64 flags,
 
 BTF_KFUNCS_START(bpf_kfunc_check_set_skb)
 BTF_ID_FLAGS(func, bpf_dynptr_from_skb, KF_TRUSTED_ARGS)
+BTF_ID_FLAGS(func, bpf_skb_checksum)
 BTF_KFUNCS_END(bpf_kfunc_check_set_skb)
 
 BTF_KFUNCS_START(bpf_kfunc_check_set_xdp)
 BTF_ID_FLAGS(func, bpf_dynptr_from_xdp)
+BTF_ID_FLAGS(func, bpf_xdp_checksum)
 BTF_KFUNCS_END(bpf_kfunc_check_set_xdp)
 
 BTF_KFUNCS_START(bpf_kfunc_check_set_sock_addr)
-- 
2.52.0

